### DISCUSSION TITLE
Improve heatmap performance using GPU rendering

### DISCUSSION THREAD

**Discussion by marcizhu** [epezent/implot#219] [Ideas]
I've been thinking about this for a while and I think we could greatly improve the performance of heatmaps if we change the way the rendering works. If I'm not mistaken, right now the CPU computes the color for each pixel, saving that data in a 2D array of some sort and then it sends to the GPU a bunch of quads to render, each one with its color and coordinates. This approach is quite effective, fast enough for most uses but it uses the CPU when in reality we could do 90% of that work on the GPU.

This is what I suggest: each colormap (the 1D array of colors indicating the range of colors we should use to render the colormap) will be stored on the GPU as a `texture1D`. The input data (the raw 2D array of points) will be a `texture2D` of `float`/`int` (as desired, probably `float`), and using a pretty basic fragment shader we render the heatmap onto an OpenGL framebuffer. Finally, instead of rendering all the quads as we do now, we just need to render a single quad and use that framebuffer as a texture. As far as I know, implot does already support rendering images on plots, so this change shouldn't be too hard to implement.

The proposed solution has quite some benefits, mainly:
1. Reduced number of vertices for heatmaps. As far as I know, for a 256x256 heatmap, we now need 256x256x3x2 vertices. That's ~393k vertices! The new approach only needs 4 vertices, but requires an additional framebuffer of size 256x256.
2. Because we are passing the raw data to the GPU, we can let OpenGL normalize it for us (so no need to calculate the minimum and maximum on the CPU) and also the color selection and interpolation will be done on the GPU, improving performance and removing a lot of work from the CPU.
3. Storing the colormap as a 1D texture on the GPU allows the user to mix and match formats for the colormap, so no more having to choose between `uint32_t` RGBA or four `float32` for colors.
4. Reduced memory usage. Now, colormaps are interpolated and stored in memory to improve CPU performance (thus using a constant memory access time instead of a more expensive linear interpolation for each pixel). By using the GPU, we don't need to store this array, and instead we can rely on the GPU for interpolation (maybe?).
5. Because we are rendering a texture, we can allow the user to choose between a linear or nearest filter. In some cases it might be desired to present a smooth heatmap without having to use thousands and thousand of points, while in other cases it might be desired to render the heatmap as is, without any interpolation.

However, I know this change has its own drawbacks. For example:
1. This solution is only for OpenGL. Supporting other backends (such as DirectX) would require a similar implementation for the desired platform. As a small fix, we could "fall back" to the current approach on unsupported platforms.
2. For small heatmaps, probably the overhead of an additional draw call isn't worth it. Real-world testing will be needed.
3. We would need to support loading & basic handling of shaders and framebuffers. Even though the shader needed for this is a one-liner (it would be something like `pixel = texture(colormap, mix(0.0f, 1.0f, texture(heatmap_data, coords)));`) we still need to handle creation, destruction and binding of shaders and framebuffers. This _con_ is not so bad though, as it would open the door to other similar optimizations for other kinds of plots, or even adding post-processing effects if the user wants to.
4. Transformations such as `Lin-Log`, `Log-Lin` or `Log-Log` must be implemented in the shader, and selected through a shader uniform (probably a couple of booleans, each one specifying if the X or Y axis is logarithmic or not).

As usual I'm open to any feedback, discussion or proposal. Maybe this is a dumb idea, but maybe it would allow us to draw massive heatmaps with close to zero CPU usage, less memory and fully taking advantage of the GPU rendering capabilities :D
Let me know what you guys think! 😄

**Comment by epezent** [comment]
@marcizhu , in general I think your idea is quite good, and I am open to having backend-specific, shader-accelerated support for our plotting functions so long as we maintain a pure ImGui fallback. Similarly, GPU accelerations could be made on thick line triangulation as well (#185 is tangentially related). I think the logical approach is to start slow, and try to build a few examples that work in OpenGL and then **maybe** we can explore other backends in the future. 

I'd love to explore this together, so let's keep the discussion going. I've been meaning to explore custom ImGui draw commands, as I think this is our path forward to inserting platform specific drawing code. Are you at all familiar with the process?

**Reply by marcizhu** [reply]
Yesterday I had some spare time and I made a small proof-of-concept in pure OpenGL and Qt for the windowing framework. This is the result:

<img width="752" alt="demo1" src="https://user-images.githubusercontent.com/6199781/115382559-21aca280-a1d5-11eb-9e80-c96ee7bbfce5.png">
And another sample with lots of data:
<img width="752" alt="demo2" src="https://user-images.githubusercontent.com/6199781/115382624-31c48200-a1d5-11eb-9694-c59b26b076d2.png">

To be honest I have never worked with custom ImGui commands, but if the documentation is good enough, it shouldn't be too hard :)

I think we can either go with the ImGui custom commands route, or just make an OpenGL backend for heatmaps. I don't know about what would be necessary for the thick line triangulation, but in this case we just need to create an image (a framebuffer object indeed) containing the rendered heatmap data. It should be easy to do both using pure, 'raw' OpenGL and using ImGui draw commands (I guess, never worked with them). It might be wise to plan ahead and use whatever will make it easier to add more shader-accelerated plots or whatever makes it easier to port the code to new platforms. I'll leave this decision up to you :)

**Reply by marcizhu** [reply]
@epezent any updates or thoughts on this? Next week I should have some free time so I could start working on this if we had a clear plan on how to implement this feature.

**Reply by epezent** [reply]
No updates. I've been just a tad busy starting a new job. 😅

**Reply by marcizhu** [reply]
No problem, first things first! I've been a bit busy with university lately so not much progress on my side either... 😅
I hope to get things going in a couple of weeks or so when I finish this semester.

**Reply by marcizhu** [reply]
@epezent a bit of a progress update: I'm now able to render a 2D matrix of floats onto a texture using a colormap, and I've got that working on top of ImGui. This image shows a simple 10x10 heatmap of random data and the "Jet" colormap rendered to a 10x10 texture:

<img width="816" alt="image" src="https://user-images.githubusercontent.com/6199781/119237389-d7547380-bb3c-11eb-9bdb-6678c2574e00.png">

My plan now is to cleanup the code and generalize it so that it can render any arbitrary number of heatmaps. After that, I think it will be easy to add that to Implot. The only issue I have right now is that I don't really know how to deal with lin-log, log-lin or log-log heatmaps. Do you know if there is any way to render an image using a custom shader from ImGui?

I plan on running some benchmarks soon, but for now, rendering a 1000x1000 heatmap only uses 15% of CPU and 19% of GPU at 60 FPS constantly on my MacBook 2018. I remember Implot had some trouble rendering a 1024x256 heatmap with the current approach so this seems definitely to be an improvement in speed! 😄

**Reply by epezent** [reply]
This is exciting! I'm looking forward to seeing more. Did you find a way to render to texture in a platform agnostic way, or is this OpenGL specific? If the latter I'm thinking we could have an optional implot_gl.cpp file and a few compile time options to set everything up correctly.

**Reply by marcizhu** [reply]
This is OpenGL-specific. I was already planning on adding a new file with some `#ifdef`/`#endif`s that wrap the entire file so that it only gets compiled (and thus used) if some macro is defined. That way Implot will use the CPU-based approach on all platforms unless there is a specific backend for that platform and the user wants to enable it.

I think I can have a PR ready in a week or two if that's okay :D

**Reply by epezent** [reply]
Absolutely! Sounds good.

**Reply by marcizhu** [reply]
@epezent I'm sorry but unfortunately this will be delayed for some days. I had some issues implementing some stuff and now I'm busy with my final uni exams. I just wanted to let you know that I haven't forgotten about this. Here's a sneak peak of what I was able to achieve so far:

https://user-images.githubusercontent.com/6199781/121084959-92d60280-c7e1-11eb-9f2c-00d19b44696f.mp4

BTW, I'm having some trouble implementing the `Log-Log`, `Lin-Log` and `Log-Lin` transformations to the axis. How does that work? Right now I just have a number on X and Y from (0,0) (lower left cornet) to (1,1) (upper right corner) and I don't know exactly how to transform that lineal scale to a logarithmic one. Any suggestions?

Thanks for your time & patience :D

**Reply by epezent** [reply]
No worries Marc. I'm swamped right now too :) 

This is really impressive work so far. I'm excited to see how it will fit into ImPlot!

Regarding the transformations -- there are 4 "Transformers" that take in plot points and output pixel coordinates. `LinLin` is regular linear scale for both x and y. LogLin is logarithmic on x, and linear on y. And so on. The math involved can be found in the transformer objects here:
https://github.com/epezent/implot/blob/3109940b46837d5c71b2d273d585bd21a88e7b81/implot_items.cpp#L421-L472

You can see that they need the size of the plot in plot coordinates and pixel coordinates, as well as a few cached computations (e.g. gp.Mx, gp.My, gp.LogDenX, gp.LogDenY). Those get pre-computed in 'BeginPlot` since they won't change once items begin. 

I think you'd need to write a fragment shader that takes in these values as uniforms to achieve the scaling. For now though, I would say it's a rather niche thing to do any lograthmic scaling on a heatmap. If it does happen, we can default back to the current method. I think we should focus on integrating what you have so far.

**Reply by marcizhu** [reply]
Yeah, I agree with you. For now I think I will focus on integrating my work with ImPlot and once that is done maybe I can try to implement other features such as the logarithmic density and such. Thanks for the details though, it will be useful once I begin working on that :D

**Reply by marcizhu** [reply]
@epezent I got it working finally!! It was easier than I thought initially. I will do some cleanup now, commit everything and open a PR as soon as possible. There are some improvements we can do to this and some things left to implement but the basics are implemented and working. As a small performance test, I ran your spectrogram demo on a small Linux computer with integrated GPU. It used to take ~45ms per frame (22 FPS or so) and required 12 draw calls to draw that heatmap. It now renders the entire heatmap in a single draw call and the FPS are capped at 60 due to VSync.

I will do more benchmarks and let you know the performance gains of this on the PR! 😄 

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/6199781/122669864-7f527080-d1bf-11eb-81c2-42ce22f3067f.png">

**Reply by epezent** [reply]
AWESOME!!! Can't wait to see this PR! I'd like to see some numbers without vsync on, and some side by side comparisons with the current implementation. 

Congrats. This is great work.

**Reply by marcizhu** [reply]
Thank you so much! I think I will have the PR ready in like an hour or two at most.
